src/shadow-catcher.ts

```diff
@@ -0,0 +1,132 @@
+import {
+    BLEND_NORMAL,
+    SHADOW_VSM16 as SHADOW_TYPE,
+    SHADOWUPDATE_REALTIME as SHADOWUPDATE,
+
+    AppBase,
+    BoundingBox,
+    Camera,
+    Entity,
+    Layer,
+    MeshInstance,
+    RenderComponent,
+    StandardMaterial
+} from 'playcanvas';
+
+const endPS = `
+    litShaderArgs.opacity = mix(light0_shadowIntensity, 0.0, shadow0);
+    gl_FragColor.rgb = vec3(0.0);
+    // gl_FragColor.rgb = vec3(mix(light0_shadowIntensity, 0.0, shadow0));
+`;
+
+class ShadowCatcher {
+    layer: Layer;
+    material: StandardMaterial;
+    plane: Entity;
+    light: Entity;
+    sceneRoot: Entity;
+    camera: Camera;
+
+    constructor(app: AppBase, camera: Camera, parent: Entity, sceneRoot: Entity) {
+        // create and add the shadow layer
+        this.layer = new Layer({
+            name: 'Shadow Layer'
+        });
+
+        const layers = app.scene.layers;
+        const worldLayer = layers.getLayerByName('World');
+        const idx = layers.getTransparentIndex(worldLayer);
+        layers.insert(this.layer, idx + 1);
+
+        // create shadow catcher material
+        this.material = new StandardMaterial();
+        this.material.useSkybox = false;
+        this.material.blendType = BLEND_NORMAL;
```

Comment:
In the WebXR Starter Kit, it uses BLEND_PREMULTIPLIED. Which is right?

Reply:
If shadow is black then they're equivalent.